File: tests/fixtures/v2-filesystem-flow/positive/FilesystemFlow.cs
class FilesystemFlow
{
    string Direct(dynamic Request)
    {
        return File.ReadAllText(Request.Query["direct"]);
    }

    void Propagated(dynamic Request, byte[] content)
    {
        var requested = Request.Query["propagated"];
        var forwardedValue = requested;
        File.WriteAllBytes(forwardedValue, content);
    }

    string Canonicalized(dynamic Request)
    {
        var canonical = Path.GetFullPath(Request.Query["protected"]);
        return File.ReadAllText(canonical);
    }

    bool Contained(string candidate, string root)
    {
        return Path.GetFullPath(candidate).StartsWith(Path.GetFullPath(root), StringComparison.Ordinal);
    }
}

File: tests/fixtures/v2-filesystem-flow/positive/filesystem-flow.js
function direct(req, fs) {
  return fs.readFileSync(req.query.direct)
}

function propagated(req, fs, content) {
  const requested = req.query.propagated
  const alias = requested
  return fs.writeFileSync(alias, content)
}

function canonicalized(req, fs, path) {
  const canonical = path.resolve(req.query.protected)
  return fs.readFileSync(canonical)
}

function contained(candidate, root, path) {
  return !path.relative(root, candidate).startsWith("..")
}
